sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -634,57 +619,93 @@ pi_result piPlatformsGet(pi_uint32 NumEntries, pi_platform *Platforms,
+          CommandListCacheSize ? std::atoi(CommandListCacheSize) : 20000;
+    } catch (std::exception const &) {
+      zePrint("SYCL_PI_LEVEL0_MAX_COMMAND_LIST_CACHE: invalid value provided, "
+              "default set.\n");
+      CommandListCacheSizeValue = 20000;
+    }
+    return CommandListCacheSizeValue;
+  })();
+
+  try {
+    // Cache pi_platforms for reuse in the future
+    // It solves two problems;
+    // 1. sycl::device equality issue; we always return the same pi_device.
+    // 2. performance; we can save time by immediately return from cache.
+    //
+    // Note: The memory for "PiPlatformsCache" and "PiPlatformsCacheMutex" is
+    // intentionally leaked because the application may call into the SYCL
+    // runtime from a global destructor, and such a call could eventually
+    // access these variables. Therefore, there is no safe time when
+    // "PiPlatformsCache" and "PiPlatformsCacheMutex" could be deleted.
+    static auto PiPlatformsCache = new std::vector<pi_platform>;
+    static auto PiPlatformsCacheMutex = new std::mutex;
+
+    std::lock_guard<std::mutex> Lock(*PiPlatformsCacheMutex);
+    for (const pi_platform CachedPlatform : *PiPlatformsCache) {
+      if (CachedPlatform->ZeDriver == ZeDriver) {
+        Platform[0] = CachedPlatform;
         return PI_SUCCESS;
       }
     }
 
-    try {
-      // TODO: figure out how/when to release this memory
-      *Platforms = new _pi_platform(ZeDriver);
-
-      // Cache driver properties
-      ze_driver_properties_t ZeDriverProperties;
-      ZE_CALL(zeDriverGetProperties(ZeDriver, &ZeDriverProperties));
-      uint32_t ZeDriverVersion = ZeDriverProperties.driverVersion;
-      // Intel Level-Zero GPU driver stores version as:
-      // | 31 - 24 | 23 - 16 | 15 - 0 |
-      // |  Major  |  Minor  | Build  |
-      std::string VersionMajor =
-          std::to_string((ZeDriverVersion & 0xFF000000) >> 24);
-      std::string VersionMinor =
-          std::to_string((ZeDriverVersion & 0x00FF0000) >> 16);
-      std::string VersionBuild = std::to_string(ZeDriverVersion & 0x0000FFFF);
-      Platforms[0]->ZeDriverVersion = VersionMajor + std::string(".") +
-                                      VersionMinor + std::string(".") +
-                                      VersionBuild;
-
-      ze_api_version_t ZeApiVersion;
-      ZE_CALL(zeDriverGetApiVersion(ZeDriver, &ZeApiVersion));
-      Platforms[0]->ZeDriverApiVersion =
-          std::to_string(ZE_MAJOR_VERSION(ZeApiVersion)) + std::string(".") +
-          std::to_string(ZE_MINOR_VERSION(ZeApiVersion));
-
-      // save a copy in the cache for future uses
-      PiPlatformsCache.push_back(Platforms[0]);
-      Platforms[0]->ZeMaxCommandListCache = CommandListCacheSizeValue;
-    } catch (const std::bad_alloc &) {
-      return PI_OUT_OF_HOST_MEMORY;
-    } catch (...) {
-      return PI_ERROR_UNKNOWN;
-    }
+    // TODO: figure out how/when to release this memory
+    *Platform = new _pi_platform(ZeDriver);
```

Comment:
If you do not need to release:
```suggestion
	static _pi_platform ZePlatform { ZeDriver };
    *Platform = &ZePlatform;
```
New rule: every time you use `new/delete/malloc/free` you have to offer some croissants to your colleagues. :-)

Reply:
We cannot do this because of the same destructor problem discussed in the comment above.  I recommend removing the TODO comment because we do NOT want to find a way to release this memory.

P.S.: My secret goal is to get as many croissants as I can from my colleagues!

Reply:
Well, I think we would prefer to eventually perform SYCL RT tear down predictably after user program destruction is completed. At this point, I agree, leaking 8 bytes here seems like the best choice currently.

Reply:
Understood, I will leave this line as it is then.